### PR TITLE
Emacs packages update: 2019-04-08

### PR DESCRIPTION
**Pull request by rasendubi**
###### Motivation for this change
Update elpa-packages, melpa-packages, melpa-stable-packages.

###### Things done

Tested all sets with nix-instantiate. Also, running Emacs from this update.

org-packages fail to update due to https://github.com/ttuegel/emacs2nix/issues/49

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cc @ttuegel

### CODE REVIEW

File: pkgs/applications/editors/emacs-modes/elpa-packages.nix [hunk 1]
```diff
@@ -45,6 +45,7 @@ self:
       stream = markBroken super.stream; # requires emacs-25
       cl-lib = null; # builtin
       tle = null; # builtin
+      advice = null; # builtin
```
Comment (rasendubi) [inline pkgs/applications/editors/emacs-modes/elpa-packages.nix]: Required by `guess-language`: http://git.savannah.gnu.org/cgit/emacs/elpa.git/commit/?h=externals/guess-language&id=32b3824d2c12ae6b3001c15018c67209c53f0fa7

File: pkgs/applications/editors/emacs-modes/melpa-packages.nix [hunk 2]
```diff
@@ -133,11 +133,7 @@ self:
       maxframe = markBroken super.maxframe;
 
       magit =
-        (super.magit.override {
-          # version of magit-popup needs to match magit
-          # https://github.com/magit/magit/issues/3286
-          inherit (self.melpaPackages) magit-popup;
-        }).overrideAttrs (attrs: {
+        super.magit.overrideAttrs (attrs: {
```
Comment (rasendubi) [inline pkgs/applications/editors/emacs-modes/melpa-packages.nix]: magit no longer uses magit-popup

**Comment by matthewbauer** [comment]
Thanks for doing this! I’m wanting to automate this process but haven’t gotten around to it yet. Usually someone comes around every month or so.

**Comment by sondr3** [comment]
@matthewbauer That would be awesome, if you have any pointers I could possibly maybe help a little 😃

**Comment by bendlas** [comment]
Thanks!

How did you get it to generate? I've failed doing so in my last attempt.

Incidentally, I've just finished a rewrite of emacs2nix, for melpa-unstable (for now). I'll link the PR.

**Comment by rycee** [comment]
Very nice! I'm wondering whether it is considered poor form to backport these packages to NixOS stable? The latest generation in 19.03 seem to be from 2019-01-03, which is quite old…

**Comment by rasendubi** [comment]
Herwig Hochleitner <notifications@github.com> writes:

> Thanks!
>
> How did you get it to generate? I've failed doing so in my last attempt.

I just followed the instructions in emacs2nix repo. It took quite some
time to run though.

How exactly it failed for you?

**Comment by rasendubi** [comment]
Robert Helgesson <notifications@github.com> writes:

> Very nice! I'm wondering whether it is considered poor form to backport these packages to NixOS stable? The latest generation in 19.03 seem to be from 2019-01-03, which is quite old…

I am not sure but I would rather not do that. The update is potentially non-backward-compatible. e.g., magit popups has been rewritten to transient which broke all my custom-defined magit commands.

**Comment by bendlas** [comment]
> How exactly it failed for you?

don't remember exactly, it's been some time. something about unstable vs stable and melpa neglecting their Makefile in favor of a Dockerfile

> It took quite some time to run though.

That's mainly because the updater re-runs the elpa builder for producing `archive.json` and `recipes.json`. Also, indexing the recipes takes a bit ..

> The update is potentially non-backward-compatible

we should probably still backport melpa (non - stable) .. when using that, the expectation of stability takes a back-seat to the expectation of recency.

File: pkgs/applications/editors/emacs-modes/melpa-packages.nix (hunk 2 above)
Comment (bendlas) [reply pkgs/applications/editors/emacs-modes/melpa-packages.nix]: after the update, magit from elpa would no longer start, complaining, that it didn't find magit-popup. It went back to normal, after I manually installed magit-popup

magit from elpaPackages seemed to have a similar proplem: starting, but being broken, also complaining about absence of magit-popup ...

**Comment by rasendubi** [comment]
Could you clarify?

1. There is no magit in elpaPackages.
2. The one in melpaStablePackages still has this work-around applied.
3. The one in melpaPackages is running fine for me. (If I revert this diff, I get an evaluation error.)
   I left a comment above—magit no longer uses magit-popup, so it must not be passed in here.

File: pkgs/applications/editors/emacs-modes/melpa-packages.nix (hunk 2 above)
Comment (rasendubi) [reply pkgs/applications/editors/emacs-modes/melpa-packages.nix]: Could you clarify?

1. There is no magit in elpaPackages.
2. The one in melpaStablePackages still has this work-around applied.
3. The one in melpaPackages is running fine for me. (If I revert this diff, I get an evaluation error.)
I left a comment above—magit no longer uses magit-popup, so it must not be passed in here.

**Comment by rycee** [comment]
For what it's worth, I cherry-picked this ontop of my 19.03 branch yesterday and have used the resulting Emacs (~100 packages) all day today without any issues. I like the new magit and lsp-java versions 🙂

File: pkgs/applications/editors/emacs-modes/melpa-packages.nix (hunk 2 above)
Comment (bendlas) [reply pkgs/applications/editors/emacs-modes/melpa-packages.nix]: Regular emacsPackages then, sorry. Here is a [before](https://github.com/bendlas/nixos-config/blob/e8b9d435bda3d95110df14f93498a2c9c92ba47f/emacs.nix) and [after](https://github.com/bendlas/nixos-config/blob/e2032ec30b6e23300837c3ace04882f53f80a3dd/emacs.nix) of my emacs config.

Without `magit-popup`, it fails. I'm on `magit-20190405.1004`

Comment (bendlas) [reply pkgs/applications/editors/emacs-modes/melpa-packages.nix]: Hm, can't reproduce it with a clean config
```
#!/bin/sh

nix-shell --pure -p git "
  with import ./. { config = { }; };
  emacsWithPackages (epkgs: [ epkgs.magit ])
" --run "HOME=/tmp emacs"
```
so it's probably fine. I'll check back, if it's something we can fix ...

Comment (bendlas) [reply pkgs/applications/editors/emacs-modes/melpa-packages.nix]: @rasendubi I found out: it's `magithub`, that still needs `magit-popup`

Comment (rasendubi) [reply pkgs/applications/editors/emacs-modes/melpa-packages.nix]: I believe this is an upstream issue.

magithub neither declares `magit-popup` as a dependency nor `(require 'magit-popup)` directly but rather relies on magit to do this.
And magit dropped `magit-popup` from dependencies now.

You might be able to reproduce the issue without Nix infrastructure using package.el: install magit and magithub (magit-popup won't be installed and you'll get the same issue).

If you need a workaround, simply pulling `magit-popup` into scope seems to fix the issue for me.
i.e., the following works fine:
```sh
nix-shell --pure -p git "
  with import ./. { config = { }; };
  emacsWithPackages (epkgs: [ epkgs.magithub epkgs.magit-popup ])
" --run "HOME=/tmp emacs"
```

**Comment by rasendubi** [comment]
Is anyone willing to merge?

**Comment by rycee** [comment]
@rasendubi Done 🙂

**Comment by rasendubi** [comment]
Thank you!

**Comment by rasendubi** [comment]
Everyone interested is welcome to review/test the next Emacs packages update: https://github.com/NixOS/nixpkgs/pull/62784

**Pull request merged**